rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs

```diff
@@ -192,92 +366,143 @@ mod test {
     }
 
     #[test]
-    fn test_throughput_log_calculate_throughput_4() {
-        let (throughput_logs, now) = build_throughput_log(1000, Duration::from_millis(100), 12);
-
-        let throughput = throughput_logs
-            .calculate_throughput(now, Duration::from_secs(1))
-            .unwrap();
-        let expected_throughput = 12.0 * 10.0;
+    fn no_polling() {
+        let start = SystemTime::UNIX_EPOCH;
+        let mut logs = ThroughputLogs::new(Duration::from_secs(1), start);
+        let report = logs.report(start + Duration::from_secs(2));
+        assert_eq!(ThroughputReport::NoPolling, report);
+    }
 
-        assert_eq!(expected_throughput, throughput.bytes_per_second());
+    // Transferred bytes MUST take priority over pending
+    #[test]
+    fn mixed_bag_mostly_pending() {
+        let start = SystemTime::UNIX_EPOCH;
+        let mut logs = ThroughputLogs::new(Duration::from_secs(1), start);
+
+        logs.push_bytes_transferred(start + Duration::from_millis(50), 10);
+        logs.push_pending(start + Duration::from_millis(150));
+        logs.push_pending(start + Duration::from_millis(250));
+        logs.push_bytes_transferred(start + Duration::from_millis(350), 10);
+        logs.push_pending(start + Duration::from_millis(450));
+        logs.push_pending(start + Duration::from_millis(650));
+        logs.push_pending(start + Duration::from_millis(750));
+        logs.push_pending(start + Duration::from_millis(850));
+
+        let report = logs.report(start + Duration::from_millis(999));
+        assert_eq!(
+            ThroughputReport::Transferred(Throughput::new_bytes_per_second(20)),
+            report
+        );
     }
 
     #[test]
-    fn test_throughput_followed_by_0() {
-        let tick = Duration::from_millis(100);
-        let (mut throughput_logs, now) = build_throughput_log(1000, tick, 12);
-        let throughput = throughput_logs
-            .calculate_throughput(now, Duration::from_secs(1))
-            .unwrap();
-        let expected_throughput = 12.0 * 10.0;
-
-        assert_eq!(expected_throughput, throughput.bytes_per_second());
-        throughput_logs.push((now + tick, 0));
-        let throughput = throughput_logs
-            .calculate_throughput(now + tick, Duration::from_secs(1))
-            .unwrap();
-        assert_eq!(108.0, throughput.bytes_per_second());
+    fn mixed_bag_mostly_pending_no_transferred() {
+        let start = SystemTime::UNIX_EPOCH;
+        let mut logs = ThroughputLogs::new(Duration::from_secs(1), start);
+
+        logs.push_pending(start + Duration::from_millis(50));
+        logs.push_pending(start + Duration::from_millis(150));
+        logs.push_pending(start + Duration::from_millis(250));
+        logs.push_pending(start + Duration::from_millis(450));
+        logs.push_pending(start + Duration::from_millis(650));
+        logs.push_pending(start + Duration::from_millis(750));
+        logs.push_pending(start + Duration::from_millis(850));
```

Comment:
alternatively,
```suggestion
        for i in (50..=850).step_by(100) {
                logs.push_pending(start + Duration::from_millis(i));
        }
```

Reply:
I didn't test this code so don't hit that commit button

Reply:
350 is skipped, so it's not quite equivalent. Perhaps it deserves a comment.